**<h1> Accelerating Insights: A Data-Driven Look at Vehicle Fuel Efficiency and Design </h1>**
#### Done By: Lamel Kekana

<div align="center" style=" font-size: 30%; text-align: center; margin: 0 auto">
<img src="https://cdn.thezebra.com/zfront/media/production/images/8xzyovqA.width-800.jpg" width="800" height="400" />
</div>

---

<h3 id="cont">Table of Contents</h1>

<a href="#BC"> Background Context</a>

<a href="#one">1. Importing Packages</a>

<a href="#two">2. Data Collection and Description</a>

<a href="#three">3. Data Loading and Inspection </a>

<a href="#four">4. Data Cleaning and Filtering</a>

<a href="#five">5. Exploratory Data Analysis (EDA)</a>

<a href="#six">6. Modeling </a>

<a href="#seven">7. Evaluation and Validation</a>

<a href="#eight">8. Final Model</a>

<a href="#nine">9. Conclusion and Future Work</a>

<a href="#ten">10. References</a>

---

## **Background Context**
<a id="BC"></a>
<a href=#cont>Back to Table of Contents</a>


#### __*Introduction*__


This dataset provides a comprehensive view of vehicle specifications and performance metrics, offering valuable insights into fuel efficiency, engine characteristics, and design trends across various car models and manufacturers. It includes information about fuel consumption in city, highway, and combined driving conditions, alongside key attributes like engine size, transmission type, and drivetrain configuration.

The dataset spans multiple vehicle classes, from compact cars to SUVs, and includes models from well-known manufacturers, with data ranging from older to newer models. By analyzing this data, we aim to uncover patterns and relationships that influence fuel efficiency, such as the impact of engine displacement, transmission type, etc on miles per gallon (mpg). Additionally, it can provide insights into trends over time, such as improvements in fuel efficiency or shifts in consumer preferences for vehicle types.

#### __*Problem Statement*__

In today’s world, fuel efficiency has become a top priority for customers who are looking to save on fuel costs and minimize their environmental footprint. The challenge lies in identifying the most fuel-efficient car models across various vehicle classes while uncovering the key factors that impact fuel economy. By tackling this issue, we can provide valuable insights to both manufacturers and consumers, helping them understand how different features, like engine size,car size and transmission affect fuel efficiency. This understanding can guide manufacturers in designing more efficient vehicles and empower consumers to make smarter, more eco-friendly purchasing decisions.

#### __*Objectives*__

- __Identify Fuel-Efficient Vehicles__

    - The goal is to identify the standout car models that excel in fuel efficiency. By ranking vehicles based on their combination MPG, the most eco-friendly options within each vehicle class can be uncovered, providing valuable insights for consumers looking to make smarter, greener choices..

- __Analyze Key Factors Influencing Fuel Efficiency__

    - This analysis will explore how key factors—such as engine displacement, the number of cylinders, drivetrain, transmission type, and fuel type—affect fuel efficiency. Understanding these elements will reveal the underlying factors that contribute to better fuel economy.

- __Class-Based Analysis__

    - Each vehicle class has its own efficiency leaders. By comparing the fuel efficiency of different types of vehicles, such as midsize, large cars, the top performers within each category will be highlighted, helping consumers choose the best option for their needs.

- __Explore Yearly Trends__

    - Examining data from 2014 to 2024, this analysis will track the progress of fuel efficiency, uncovering how technological advancements and design improvements have contributed to more fuel-efficient vehicles, and offering insights into the future of eco-friendly driving.   

---
## **Importing Packages**
<a id="one"></a>
<a href=#cont>Back to Table of Contents</a>



In [44]:
# importing packages
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

---

## **Data Collection and Description**
<a id="two"></a>
<a href=#cont>Back to Table of Contents</a>


#### __*Source*__

The dataset was collected from Kaggle, an online platform that hosts a variety of datasets shared by individuals, organizations, and data science communities. Kaggle datasets typically come from real-world scenarios and are often used for machine learning competitions, research, and educational purposes.

#### __*Description*__

Key Features:
- **city_mpg:** Fuel efficiency in miles per gallon (mpg) for city driving, crucial for urban driving analysis.
- **highway_mpg:** Fuel efficiency in mpg for highway driving, ideal for comparing long-distance performance.
- **combination_mpg:**** The combined fuel efficiency for both city and highway driving, a critical metric for overall vehicle performance.
- **cylinders & displacement:**** Engine specifications that help understand a car's power and fuel consumption.
- **drive:** The type of drivetrain (e.g., FWD, AWD), key for determining the car's handling and terrain suitability.
- **fuel_type:** The type of fuel (e.g., gasoline, electric), important for environmental impact studies.
- **make & model:** Car manufacturer and model, enabling brand-specific analysis and comparison.
- **year:** The production year of the vehicle, useful for tracking trends and technological improvements over time.

---
## **Data Loading and Inspection**
<a id="three"></a>
<a href=#cont>Back to Table of Contents</a>


The first step in the process involves loading the dataset into the analysis environment, ensuring that all necessary data is properly imported for further analysis

In [45]:
# Loading the data 
df = pd.read_csv('car_data.csv')

In [46]:
# Overview of the dataset
df.head()

,city_mpg,class,combination_mpg,cylinders,displacement,drive,fuel_type,highway_mpg,make,model,transmission,year
0,25,midsize car,29,4.0,2.5,fwd,gas,36,mazda,6,m,2014
1,26,midsize car,30,4.0,2.5,fwd,gas,37,mazda,6,a,2014
2,25,small sport utility vehicle,27,4.0,2.5,fwd,gas,31,mazda,cx-5 2wd,a,2014
3,26,small sport utility vehicle,29,4.0,2.0,fwd,gas,34,mazda,cx-5 2wd,m,2014
4,26,small sport utility vehicle,28,4.0,2.0,fwd,gas,32,mazda,cx-5 2wd,a,2014


In [47]:
# overview of structure and datatypes
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 550 entries, 0 to 549
Data columns (total 12 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   city_mpg         550 non-null    int64  
 1   class            550 non-null    object 
 2   combination_mpg  550 non-null    int64  
 3   cylinders        548 non-null    float64
 4   displacement     548 non-null    float64
 5   drive            550 non-null    object 
 6   fuel_type        550 non-null    object 
 7   highway_mpg      550 non-null    int64  
 8   make             550 non-null    object 
 9   model            550 non-null    object 
 10  transmission     550 non-null    object 
 11  year             550 non-null    int64  
dtypes: float64(2), int64(4), object(6)
memory usage: 51.7+ KB


The dataset consists of 550 entries and mix of numeric and object datatype, but some columns appear to contain missing values

In [48]:
# describing the dataset statistics
df.describe()

,city_mpg,combination_mpg,cylinders,displacement,highway_mpg,year
count,550.000000,550.000000,548.000000,548.000000,550.000000,550.000000
mean,21.460000,24.069091,5.315693,2.931752,28.609091,2019.000000
std,8.147392,7.478369,1.759999,1.248419,6.832228,3.165156
min,11.000000,14.000000,3.000000,1.200000,18.000000,2014.000000
25%,17.000000,20.000000,4.000000,2.000000,24.000000,2016.000000
50%,20.000000,23.000000,4.000000,2.500000,28.000000,2019.000000
75%,24.000000,27.000000,6.000000,3.500000,32.000000,2022.000000
max,126.000000,112.000000,12.000000,6.800000,102.000000,2024.000000


__Fuel Economy__

- The very high miles-per-gallon (MPG) values, such as 126 MPG for city driving and 112 MPG for combined driving might be outliers or it could indicate a possible mistake in data entry or the presence of vehicles designed for exceptional fuel efficiency, such as advanced hybrid cars.

__Engine Characteristics__

- Vehicles have between 3 to 12 cylinders, with a median of 4 cylinders, suggesting most cars are small to midsize.
- Engine displacement ranges from 1.2L to 6.8L, with a median of 2.5L, indicating a mix of economy and performance vehicles.

Aside from the extreme MPG values, most of the other values in the statistical summary table appear realistic

---
## **Data Cleaning and Filtering**
<a id="four"></a>
<a href=#cont>Back to Table of Contents</a>


The data cleaning process will involve identifying and addressing missing values, duplicates, outliers, and any other inconsistencies in the dataset.

***Checking for missing values across all rows***

In [49]:
# Check for missing values across all the rows
missing_val_rows = df.isnull().any(axis=1).sum()

# Print out number of missing values
print("No of missing values across all rows: ", missing_val_rows)

No of missing values across all rows:  2


In [50]:
# view the rows with missing values
df[df.isnull().any(axis=1)]

,city_mpg,class,combination_mpg,cylinders,displacement,drive,fuel_type,highway_mpg,make,model,transmission,year
128,126,subcompact car,112,NaN,NaN,rwd,electricity,99,mitsubishi,i-miev,a,2016
175,121,subcompact car,112,NaN,NaN,rwd,electricity,102,mitsubishi,i-miev,a,2017


Wow,As suspected, when the statistical table was viewed, there were unrealistically high maximum values for city mpg and combination mpg. While the suspicion was that these might be hybrid vehicles, it turns out they are actually electric vehicles. It's no surprise to see that the missing values are in the cylinders and displacement features, as EVs use electric motors instead of traditional internal combustion engines. EVs do not have cylinders; instead, they rely on components like electric motors, batteries, and controllers to manage power delivery to the wheels. As a result, these two rows will be deleted, and other electric vehicles will be checked and removed as well, since this is a fuel analysis and not an EV analysis. Although some other vehicles in the dataset are suspected to be hybrids, they won't be deleted, as they still use fuel, and their fuel efficiency is optimized by the electric motor.

In [51]:
# drop rows with missing values
df.dropna(axis=0, how='any', inplace=True)

In [52]:
# confirm if rows with missing values were succefully dropped
missing_val_rows = df.isnull().any(axis=1).sum()

# Print out number of missing values
print("No of missing values across all rows: ", missing_val_rows)


No of missing values across all rows:  0


***Checking for missing values across all columns***

In [53]:
# check missing values across all columns
df.isnull().sum()

city_mpg           0
class              0
combination_mpg    0
cylinders          0
displacement       0
drive              0
fuel_type          0
highway_mpg        0
make               0
model              0
transmission       0
year               0
dtype: int64

***Checking if there are Electrical Vehichles in the dataset***

In [ ]:
# print out all fuel_types in dataset to see if theres Electrical Vehicales
print(df['fuel_type'].unique())

['gas' 'diesel']


    There are no other electric vehicles (EVs) in the dataset

***Check for duplicate rows***

In [55]:
# check for duplicate rows
no_duplicates = df.duplicated().sum()

# print out number of duplicate rows
print("No of duplicated rows: ", no_duplicates )

No of duplicated rows:  2


In [56]:
# view the duplicate rows
df[df.duplicated(keep=False)]

,city_mpg,class,combination_mpg,cylinders,displacement,drive,fuel_type,highway_mpg,make,model,transmission,year
116,28,small sport utility vehicle,31,4.0,1.8,fwd,gas,34,honda,hr-v 2wd,a,2016
117,28,small sport utility vehicle,31,4.0,1.8,fwd,gas,34,honda,hr-v 2wd,a,2016
118,27,small sport utility vehicle,29,4.0,1.8,awd,gas,32,honda,hr-v 4wd,a,2016
119,27,small sport utility vehicle,29,4.0,1.8,awd,gas,32,honda,hr-v 4wd,a,2016


The first occurance of each entry that was duplicated will be dropped

In [57]:
# drop duplicates
df.drop_duplicates(inplace=True)

In [58]:
# No of duplicate rows after dropping them
no_duplicates = df.duplicated().sum()

# check if duplicates were succesfully dropped
print("No of duplicated rows: ", no_duplicates )

No of duplicated rows:  0


After removing duplicate rows with identical values across all columns, additional similar entries were observed where records differed in the year column or showed slight, insignificant variations in the mpg-related columns. These discrepancies introduce redundancy in the dataset. Although the production year differs, the cars share the same specifications, with fuel economy variations of only 1-2 miles per gallon or in some cases, no change at all. To address this, only the latest year models will be retained, as they are more likely to be efficient than previous year models due to optimizations and tuning.

In [ ]:
# Identify rows that have duplicate values based on the specified subset of columns.
dup_diff_year = df[df.duplicated(subset=['displacement', 'model', 'fuel_type', 'make','transmission','cylinders'], keep=False)]# The 'keep=False' parameter ensures that all duplicate rows are marked

# Sort the identified duplicates by the specified columns for better organization and clarity.
dup_diff_year= dup_diff_year.sort_values(by=['make', 'model', 'transmission','displacement'])

# overview of filtered duplicated df
dup_diff_year.head(6)

,city_mpg,class,combination_mpg,cylinders,displacement,drive,fuel_type,highway_mpg,make,model,transmission,year
178,24,compact car,27,4.0,2.0,awd,gas,31,audi,a4 quattro,a,2017
204,24,compact car,27,4.0,2.0,awd,gas,34,audi,a4 quattro,a,2018
25,13,two seater,16,10.0,5.2,awd,gas,22,audi,r8,a,2014
176,14,two seater,17,10.0,5.2,awd,gas,22,audi,r8,a,2017
24,13,two seater,16,10.0,5.2,awd,gas,22,audi,r8 spyder,a,2014
177,14,two seater,17,10.0,5.2,awd,gas,22,audi,r8 spyder,a,2017


In [60]:
# Sort the DataFrame by 'year' in descending order (latest year first)
df = df.sort_values(by='year', ascending=False)

In [ ]:
# Drop entries that have similiar specs but only retain latest year models
df = df.drop_duplicates(subset=['displacement', 'model', 'make', 'transmission', 'cylinders'], keep='first')

df

,city_mpg,class,combination_mpg,cylinders,displacement,drive,fuel_type,highway_mpg,make,model,transmission,year
549,17,midsize car,19,8.0,4.4,awd,gas,24,bmw,alpina b8 gran coupe,a,2024
511,25,small sport utility vehicle,26,4.0,1.6,awd,gas,27,kia,seltos awd,a,2024
522,16,subcompact car,19,6.0,3.0,rwd,gas,23,bmw,m4 coupe,m,2024
521,23,subcompact car,27,4.0,2.0,awd,gas,33,bmw,430i xdrive convertible,a,2024
520,25,subcompact car,28,4.0,2.0,rwd,gas,34,bmw,430i convertible,a,2024
...,...,...,...,...,...,...,...,...,...,...,...,...
32,13,minicompact car,15,12.0,5.9,rwd,gas,19,aston martin,db9,a,2014
31,11,two seater,14,8.0,4.2,awd,gas,20,audi,r8,m,2014
30,11,two seater,14,8.0,4.2,awd,gas,20,audi,r8 spyder,m,2014
29,12,two seater,14,10.0,5.2,awd,gas,19,audi,r8,m,2014


***Check for outliers***

The combination_mpg column, which represents the combined city and highway miles per gallon (MPG), will be used to calculate outliers in the dataset.

In [62]:
# Calculate the first(Q1) and third quartile(Q3) of the 'combination_mpg' column.
Q1 = df['combination_mpg'].quantile(0.25)
Q3 = df['combination_mpg'].quantile(0.75)

# Calculate iner quartile range
IQR = Q3 - Q1

# Calculate lower bond and upper bownd
lower_bound = (Q1 - 1.5 * IQR)
upper_bound = (Q3 + 1.5 * IQR)

# Define outliers as values below lower bound or above upper bound
outliers = df[(df['combination_mpg'] < lower_bound) | (df['combination_mpg'] > upper_bound)]

In [63]:
# View the outliers
outliers

,city_mpg,class,combination_mpg,cylinders,displacement,drive,fuel_type,highway_mpg,make,model,transmission,year
477,38,small sport utility vehicle,38,4.0,1.6,awd,gas,38,kia,sportage hybrid awd,a,2023
478,42,small sport utility vehicle,43,4.0,1.6,fwd,gas,44,kia,sportage hybrid fwd,a,2023
447,38,small sport utility vehicle,38,4.0,1.6,awd,gas,38,hyundai,tucson hybrid blue,a,2022
391,55,midsize car,52,4.0,1.5,fwd,gas,49,honda,insight,a,2021
386,51,compact car,48,4.0,1.5,fwd,gas,45,honda,insight touring,a,2021
301,53,compact car,52,4.0,1.8,fwd,gas,52,toyota,corolla hybrid,a,2020
173,37,compact car,39,3.0,1.2,fwd,gas,43,mitsubishi,mirage,a,2017
186,43,compact car,42,4.0,1.5,fwd,gas,42,chevrolet,volt,a,2017


As expected, the dataset contains **hybrid** vehicles, even though **hybrid** isn't explicitly mentioned in the **fuel_type** column. Instead, it's indicated by the vehicle model names, which explains why some vehicles were flagged as outliers, due to their high value(fuel efficiency) in **combination_mpg**

One reason not to delete these hybrids is the risk involved. For example, the **Chevrolet Volt** (the second entry in the table) is a hybrid, even though the name doesn't explicitly mention "hybrid." However, the name Volt itself is a giveaway, as it refers to electricity. Similarly, the **Honda Insight Touring** (the fourth entry) is also a hybrid, which was confirmed through external research.

The challenge is that hybrids are not clearly identified in the **fuel_type** column, and some hybrids might not have been flagged as outliers at all.

For this reason, the oultiers will be **retained**. These entries could be valuable for further analysis, as they may contain important information that shouldn't be discarded.

***Check for class imbalance***

In [64]:
# check vehicle classes
df.groupby('class')['class'].count()

class
compact car                        54
large car                          13
midsize car                        36
midsize station wagon               2
minicompact car                     8
minivan                             6
small pickup truck                  7
small sport utility vehicle       123
small station wagon                 7
standard pickup truck               7
standard sport utility vehicle     34
subcompact car                     56
two seater                         37
Name: class, dtype: int64

Since the **midsize station wagon** category has only two entries, it introduces a bias due to the small sample size. To address this, the **midsize station wagon**, **small station wagon**, and **minivan** categories will be combined into a single category, creating a more balanced distribution.  

The **minivan** category is also included in this grouping because, like station wagons, minivans are designed for passenger and cargo transport, offering practicality and space efficiency. Combining these categories helps reduce class imbalance while maintaining meaningful distinctions between vehicle types.


Similarly, the **small pickup truck** and **standard pickup truck** categories will be merged into a single **pickup truck** category. Additionally, other class categories will also be consolidated into a single category. This consolidation helps reduce class imbalance, ensuring that each category has a more representative number of entries.



In [ ]:
# function to reclassificy cars 
def classify_car(car_class):
    """
    Classifies a car into broader categories to reduce class imbalance.
    
    Args:
        car_class (str): The specific car classification.
    
    Returns:
        str: The generalized car category.
    """

    # Classify cars based on similarity
    if car_class in ["small pickup truck", "standard pickup truck"]:
        return "pickup truck"
    elif car_class in ["midsize station wagon", "small station wagon","minivan"]:
        return "wagon/minivan"
    elif car_class in ["minicompact car","compact car","subcompact car"]:
        return "compact car"
    
    return car_class  # Returns the original class if no match is found


In [66]:
# apply the classify_car function on the class column
df.loc[:, 'class'] = df['class'].apply(classify_car)

In [67]:
# check for vehicle classes 
df.groupby('class')['class'].count()

class
compact car                       118
large car                          13
midsize car                        36
pickup truck                       14
small sport utility vehicle       123
standard sport utility vehicle     34
two seater                         37
wagon/minivan                      15
Name: class, dtype: int64

Although some imbalance remains, particularly in the **large car**, **pickup truck**, and **wagon/minivan** categories, which still have relatively fewer entries, this approach ensures that most categories have a minimum of at least 10 entries. While merging similar categories helps reduce extreme bias, it does not completely eliminate disparities in class distribution. Despite this limitation, the approach improves overall balance while preserving meaningful distinctions between vehicle types.



***Checking the data info after cleaning***

In [ ]:
# data information after cleaning
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 390 entries, 549 to 28
Data columns (total 12 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   city_mpg         390 non-null    int64  
 1   class            390 non-null    object 
 2   combination_mpg  390 non-null    int64  
 3   cylinders        390 non-null    float64
 4   displacement     390 non-null    float64
 5   drive            390 non-null    object 
 6   fuel_type        390 non-null    object 
 7   highway_mpg      390 non-null    int64  
 8   make             390 non-null    object 
 9   model            390 non-null    object 
 10  transmission     390 non-null    object 
 11  year             390 non-null    int64  
dtypes: float64(2), int64(4), object(6)
memory usage: 39.6+ KB


The dataset has been reduced by approximately **28.83%** (from 550 to 390 rows). While this may seem like a significant reduction, it is a worthwhile sacrifice as it eliminates redundancy and noise, resulting in a cleaner and more reliable dataset for analysis.

---

## **Exploratory Data Analysis (EDA)**
<a id="five"></a>
<a href=#cont>Back to Table of Contents</a>

* **Purpose:** Explore and visualize the data to uncover patterns, trends, and relationships.
* **Details:** Use statistics and visualizations to explore the data. This may include histograms, box plots, scatter plots, and correlation matrices. Discuss any significant findings.
---


In [ ]:
#Please use code cells to code in and do not forget to comment your code.

---

## **Modeling**
<a id="six"></a>
<a href=#cont>Back to Table of Contents</a>

* **Purpose:** Develop and train predictive or statistical models.
* **Details:** Describe the choice of models, feature selection and engineering processes, and show how the models are trained. Include code for setting up the models and explanations of the model parameters.
---


In [ ]:
#Please use code cells to code in and do not forget to comment your code.

---

## **Evaluation and Validation**
<a id="seven"></a>
<a href=#cont>Back to Table of Contents</a>

* **Purpose:** Evaluate and validate the effectiveness and accuracy of the models.
* **Details:** Present metrics used to evaluate the models, such as accuracy, precision, recall, F1-score, etc. Discuss validation techniques employed, such as cross-validation or train/test split.
---

In [ ]:
#Please use code cells to code in and do not forget to comment your code.

---

## **Final Model**
<a id="eight"></a>
<a href=#cont>Back to Table of Contents</a>

* **Purpose:** Present the final model and its performance.
* **Details:** Highlight the best-performing model and discuss its configuration, performance, and why it was chosen over others.
---


In [ ]:
#Please use code cells to code in and do not forget to comment your code.

---

## **Conclusion and Future Work**
<a id="nine"></a>
<a href=#cont>Back to Table of Contents</a>

* **Purpose:** Summarize the findings and discuss future directions.
* **Details:** Conclude with a summary of the results, insights gained, limitations of the study, and suggestions for future projects or improvements in methodology or data collection.
---


In [ ]:
#Please use code cells to code in and do not forget to comment your code.

---

## **References**
<a id="ten"></a>
<a href=#cont>Back to Table of Contents</a>

* **Purpose:** Provide citations and sources of external content.
* **Details:** List all the references and sources consulted during the project, including data sources, research papers, and documentation for tools and libraries used.
---

In [ ]:
#Please use code cells to code in and do not forget to comment your code.

## Additional Sections to Consider

* ### Appendix: 
For any additional code, detailed tables, or extended data visualizations that are supplementary to the main content.

* ### Contributors: 
If this is a group project, list the contributors and their roles or contributions to the project.
